File: tp_tda/entregables/2/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int creciente(int buildings_q, vector<pair<int, int>>& sector, vector<int>& mem) {
    int max_val = 0;
    if (buildings_q == 0)
        return 0;
    if (buildings_q == 1)
        return sector[0].second;
    for (int i = 0; i < buildings_q; ++i) {
        mem[i] = sector[i].second;
        for (int j = 0; j < i; ++j) {
            if (sector[j].first < sector[i].first && sector[i].second + mem[j] > mem[i]) {
                mem[i] = sector[i].second + mem[j];
            }
        }
        if (mem[i] > mem[max_val]) {
            max_val = i;
        }
    }
    return mem[max_val];
}

int decreciente(int buildings_q, vector<pair<int, int>>& sector, vector<int>& mem) {
    int max_val = 0;
    if (buildings_q == 0)
        return 0;
    if (buildings_q == 1)
        return sector[0].second;
    for (int i = 0; i < buildings_q; ++i) {
        mem[i] = sector[i].second;
        for (int j = 0; j < i; ++j) {
            if (sector[j].first > sector[i].first && sector[i].second + mem[j] > mem[i]) {
                mem[i] = sector[i].second + mem[j];
            }
        }
        if (mem[i] > mem[max_val]) {
            max_val = i;
        }
    }
    return mem[max_val];
}

int main() {
    int case_count;
    cin >> case_count;
    vector<vector<pair<int, int>>> mem(case_count);
    for (int i = 0; i < case_count; i++) {
        int building_q;
        cin >> building_q;
        vector<int> heights(building_q);
        vector<int> widths(building_q);
        for (int j = 0; j < building_q; j++) {
            cin >> heights[j];
        }
        for (int j = 0; j < building_q; j++) {
            cin >> widths[j];
        }
        vector<pair<int, int>> sector(building_q);
        for (int j = 0; j < building_q; j++) {
            sector[j] = make_pair(heights[j], widths[j]);
        }
        mem[i] = sector;
    }
    vector<pair<int, int>> skyline_status;
    for (int i = 0; i < case_count; ++i) {
        int q = mem[i].size();
        vector<int> cre_mem(q, 0);
        vector<int> dec_mem(q, 0);
        int cre_val = creciente(q, mem[i], cre_mem);
        int dec_val = decreciente(q, mem[i], dec_mem);
        skyline_status.push_back({cre_val, dec_val});
        if (skyline_status[i].first >= skyline_status[i].second) {
            cout << "Case " << i + 1 << ". Increasing (" << skyline_status[i].first << "). Decreasing (" << skyline_status[i].second << ")." << endl;
        } else {
            cout << "Case " << i + 1 << ". Decreasing (" << skyline_status[i].second << "). Increasing (" << skyline_status[i].first << ")." << endl;
        }
    }
    return 0;
}

File: tp_tda/entregables/2/main.py
def creciente(buildings_q, sector, mem):
    max = 0
    if buildings_q == 0:
        return 0
    if buildings_q == 1:
        return sector[0][1]
    for i in range(buildings_q):
        mem[i] = sector[i][1]
        
        for j in range(i):
            if sector[j][0] < sector[i][0] and sector[i][1] + mem[j] > mem[i]:
                mem[i] = sector[i][1] + mem[j]
        
        if mem[i] > mem[max]:
            max = i
    
    return mem[max]

def decreciente(buildings_q, sector, mem):
    max = 0
    if buildings_q == 0:
        return 0
    if buildings_q == 1:
        return sector[0][1]
    for i in range(buildings_q):
        mem[i] = sector[i][1]
        
        for j in range(i):
            if sector[j][0] > sector[i][0] and sector[i][1] + mem[j] > mem[i]:
                mem[i] = sector[i][1] + mem[j]
        
        if mem[i] > mem[max]:
            max = i
    
    return mem[max]

def main():
    case = int(input())
    mem = []
    for i in range(0, case, 1):
        building_q = int(input())
        sector = []
        building_h = input().split()
        building_w = input().split()
        for j in range (0, len(building_h), 1):
            building_h[j] = int(building_h[j])
            building_w[j] = int(building_w[j])
            buildings = (building_h[j], building_w[j])
            sector.append(buildings)
        mem.append(sector)
    skyline_status = []
    for c in range(0, case, 1):
        q = len(mem[c])
        inc_mem = [0 for _ in range(q)]
        dec_mem = [0 for _ in range(q)]
        
        skyline = (creciente(q, mem[c], inc_mem), decreciente(q, mem[c], dec_mem))
        skyline_status.append(skyline)
        
        if skyline_status[c][0] >= skyline_status[c][1]:
            res = f"Case {c+1}. Increasing ({skyline_status[c][0]}). Decreasing ({skyline_status[c][1]})."
            print(res)
        else:
            res = f"Case {c+1}. Decreasing ({skyline_status[c][1]}). Increasing ({skyline_status[c][0]})."
            print(res) 

main()